src/data/list.rs

```diff
@@ -1,45 +1,67 @@
 use crate::{applicative::Applicative, functor::Functor, functor::Pointed, monad::Monad};
 
-// pub enum List<T> {
-//     Cons { head: T, tail: Box<List<T>> },
-//     Nil,
-// }
+#[derive(Debug, PartialEq)]
+pub enum List<T> {
+    Cons { head: T, tail: Box<List<T>> },
+    Nil,
+}
 
-// impl<A> Pointed for List<A> {
-//     fn pure(t: A) -> List<A> {
-//         List::Cons {
-//             head: t,
-//             tail: Box::new(List::Nil),
-//         }
-//     }
-// }
+impl<A> Pointed for List<A> {
+    fn pure(t: A) -> List<A> {
+        List::Cons {
+            head: t,
+            tail: Box::new(List::Nil),
+        }
+    }
+}
 
-// impl<A> Functor for List<A> {
-//     type A = A;
-//     type Lifted<B> = Functor<B>;
+impl<A> Functor for List<A> {
+    type A = A;
+    type Lifted<B> = List<B>;
 
-//     fn map<A, B>(self, f: F) -> List<B>
-//     where
-//         F: Fn(A) -> B,
-//     {
-//         todo!()
-//         // match self {
-//         //     List::Cons(head, tail)
-//         // }
-//     }
-// }
+    fn map<F, B>(self, mut f: F) -> List<B>
+    where
+        F: FnMut(Self::A) -> B,
+    {
+        match self {
+            List::Cons { mut head, tail} =>
+                List::Cons { head: f(head), tail: Box::new(tail.map(f)) },
```

Comment:
This is NOT tail-recursive but we need twice loop to tail-recursive.

```rust
let go = |acc: List<B>, list: List<A>| -> List<B> {
   match list {
     List::Nil =>
       acc,
     List::Cons { mut head, tail } =>
        go(
          List::Cons { head: f(head), tail: acc },
          tail
        )
   };
```

This `go` is tail-recursive but the result is reversed so we have more 1 loop to revert the reversed result.